training/coach.py

```diff
@@ -52,42 +53,54 @@ def get_model(self) -> Kandinsky2_1:
         elif self.cfg.learnable_property == LearnableProperties.style:
             # NOTE: We specifically specify the content to avoid the model adding it to the answer
             # Currently hard-coded to match the images generated in the style mode
             question = f"What art style was used in this photo of a horse and a barn?"
         else:
             raise ValueError(f"Unknown learnable property: {self.cfg.learnable_property}")
 
-        with torch.no_grad():
-            inputs = self.blip_processor(sampled_image, question, return_tensors="pt").to("cuda", torch.float16)
-            out = self.blip_model.generate(**inputs)
-            negative_answer = self.blip_processor.decode(out[0], skip_special_tokens=True)
+        neg_classes = []
+
+        for sampled_image in sampled_images:
+            with torch.no_grad():
+                inputs = self.blip_processor(sampled_image, question, return_tensors="pt").to("cuda", torch.float16)
+                out = self.blip_model.generate(**inputs)
+                negative_answer = self.blip_processor.decode(out[0], skip_special_tokens=True)
+
+            # Remove leading a, if exists
+            negative_cls = negative_answer
+            if negative_cls.startswith("a "):
+                negative_cls = negative_cls[2:]
 
-        # Remove leading a, if exists
-        negative_cls = negative_answer
-        if negative_cls.startswith("a "):
-            negative_cls = negative_cls[2:]
+            print(f'Idendified class: "{negative_cls}"')
+            neg_classes.append(negative_cls)
 
-        print(f'Adding negative class: "{negative_cls}"')
-        return negative_cls
+        return neg_classes
 
-    def save_images(self, save_dir: Path, save_prefix: str):
+    def save_images(self, save_dir: Path, save_prefix: str, image_embs: list = None, template = None, fix_seed = True):
         if self.cfg.learnable_property == LearnableProperties.style:
             prompts = [f"a painting of a horse and a barn in a valley in the style of {self.cfg.placeholder_token}",
                        f"a painting of a dog in the style of {self.cfg.placeholder_token}",
                        f"a painting of fruit bowl in the style of {self.cfg.placeholder_token}"]
         else:
             # First prompt is the one we are going to run blip against, this specific one works great with kandinsky
-            prompts = [f"Professional high-quality art of a {self.cfg.placeholder_token}. photorealistic, 4k, HQ",
+            if template is None:
+                prompts = [f"Professional high-quality art of a {self.cfg.placeholder_token}. photorealistic, 4k, HQ",
                        f"a photo of a {self.cfg.placeholder_token}",
                        random.choice(object_templates_edits).format(a='a', token=self.cfg.placeholder_token)]
+            else:
+                prompts = [template[0].format(token=self.cfg.placeholder_token)]
```

Comment:
we generate negative image from this function
1. use our own prompts to generate negative image
2. the prompts will be align with positive image in the same iteration